File: dev-server.mjs
import http from 'node:http';
import { createReadStream, statSync } from 'node:fs';
import { extname, resolve, sep } from 'node:path';

const root = resolve(process.cwd());
const port = Number(process.env.PORT || 4180);
const mime = {
  '.css': 'text/css; charset=utf-8',
  '.html': 'text/html; charset=utf-8',
  '.js': 'text/javascript; charset=utf-8',
  '.json': 'application/json; charset=utf-8',
  '.png': 'image/png',
  '.svg': 'image/svg+xml',
  '.webmanifest': 'application/manifest+json; charset=utf-8'
};

const server = http.createServer((request, response) => {
  try {
    const pathname = decodeURIComponent(new URL(request.url || '/', 'http://127.0.0.1').pathname);
    const requested = pathname === '/' ? 'index.html' : pathname.replace(/^\/+/, '');
    const file = resolve(root, requested);
    if (file !== root && !file.startsWith(`${root}${sep}`)) throw new Error('outside root');
    if (!statSync(file).isFile()) throw new Error('not a file');
    response.writeHead(200, {
      'Content-Type': mime[extname(file).toLowerCase()] || 'application/octet-stream',
      'Cache-Control': 'no-store, no-cache, must-revalidate, max-age=0',
      'Pragma': 'no-cache',
      'X-Content-Type-Options': 'nosniff'
    });
    createReadStream(file).pipe(response);
  } catch (error) {
    response.writeHead(404, { 'Content-Type': 'text/plain; charset=utf-8', 'Cache-Control': 'no-store' });
    response.end('Not found');
  }
});

server.listen(port, '127.0.0.1', () => {
  console.log(`Second Brain OS dev server: http://127.0.0.1:${port}/`);
});

File: dev-server.py
#!/usr/bin/env python3
"""Локальный сервер для разработки.

Отличается от `python -m http.server` одним: запрещает кэширование.

Это важно именно здесь. Приложение собрано из ES-модулей, которые
подключаются по относительным путям без версии, и браузер держит их в
памяти. После правки файла в v2/src/ обычный сервер отдаёт 304, браузер
берёт старый код — и правка будто не применилась. На этом легко потерять
полчаса, отлаживая уже исправленную ошибку.

Запуск:  python dev-server.py [порт]
"""

import sys
from functools import partial
from http.server import SimpleHTTPRequestHandler, ThreadingHTTPServer


class NoCacheHandler(SimpleHTTPRequestHandler):
    def end_headers(self):
        self.send_header("Cache-Control", "no-store, no-cache, must-revalidate, max-age=0")
        self.send_header("Pragma", "no-cache")
        self.send_header("Expires", "0")
        super().end_headers()

    def log_message(self, fmt, *args):
        """Тихий лог: только ошибки, иначе консоль тонет в запросах модулей."""
        status = str(args[1]) if len(args) > 1 else ""
        if status.startswith("4") or status.startswith("5"):
            super().log_message(fmt, *args)


def main():
    port = int(sys.argv[1]) if len(sys.argv) > 1 else 8834
    handler = partial(NoCacheHandler, directory=".")
    server = ThreadingHTTPServer(("127.0.0.1", port), handler)
    print(f"Прежняя версия: http://localhost:{port}/index.html")
    print(f"Новая версия:   http://localhost:{port}/v2/index.html")
    print("Остановить — Ctrl+C")
    try:
        server.serve_forever()
    except KeyboardInterrupt:
        print("\nостановлен")


if __name__ == "__main__":
    main()
